rfcs/0077-layout-transform-padding.md

```diff
@@ -0,0 +1,2522 @@
+for its removal.
+
+
+### Buffer Annotation of Layout Transforms
+
+TODO: Should a buffer remember which layout transforms have been
+applied to it?  It would be useful for generating converters between
+logical/transformed/physical layout.  As it is, users must provide
+inputs that have the transformed layout.
+
+## Transformations/Metaschedule Primitives
+
+### Enhancement - transform_layout
+
+The `te.Stage.transform_layout` and `tir.Schedule.transform_layout`
+methods will be updated to take an additional argument `pad_value:
+Optional[Union[int, float, Callable]]`.  This provides the `value`
+field of the `BufferConstraintNode`.
+
+For buffer consumers, the buffer constraint is updated, and no further
+changes are required based on the padding value.  For buffer
+producers, the buffer constraint is updated, and an additional loop is
+added to write `pad_value` to the padding that has been introduced.
+
+```python
+# Before transforming A
+@T.prim_func
+def func(A: T.Buffer[(14,), "float32"]):
+    for i in T.serial(14):
+        A[i] = i
+
+# After applying transform_layout(lambda i: [i//4, i%4], pad_value=-1)
+@T.prim_func
+def func(A: T.Buffer[(4,4), "int32"]):
+    # This loop writes the same values, but to the new locations in
+    # `A`.
+    for i in T.serial(14):
+        A[i//4, i%4] = i
+
+    # This loop writes the padding values.  In this case, `io==3 and
+    # ii>2` is the predicate, and `-1` is the value.
+    for io,ii in T.grid(4,4):
+        if io==3 and ii>2:
+            A[io, ii] = -1
+```
+
+It is expected that the loop that writes padding may be simplified
+later.  In this case, the loop over `io` can be removed, and the range
+of the loop over `ii` can be reduced to `2 <= ii < 4`.  However, the
+default implementation should not perform these simplification yet, as
+this form is useful for [merging
+loopnests](#utility-merge-adjacent-loops) after [rewriting for
+sequential buffer
+access](#new-primitive-reorder-loops-according-to-buffer).
+
+In TE, the producer is the stage that outputs the transformed tensor.
+In TIR, the producer is the block that writes to all values of the
+pre-transformation tensor.
+
+
+
+### New Primitive - Add buffer constraint
+
+Similar to `Schedule.set_axis_separators`, this adds an annotation to
+an existing buffer, and can be used independently of
+`transform_layout`.  This can be useful for hardware that provides a
+default value for out-of-bounds reads (e.g. texture memory clamping on
+a GPU).
+
+### New Primitive - Reorder Loops According to Buffer
+
+By default in S-TIR, `transform_layout` modifies the underlying layout
+of a buffer, but does not re-order loops that iterate over the buffer.
+A new S-TIR transformation `Schedule.sequential_buffer_access` should
+be introduced, which rewrites iteration loops according to the access
+pattern of a buffer.
+
+```python
+# Original function
+@T.prim_func
+def func(A: T.Buffer[(16,), "int32"]):
+    with T.block('compute'):
```

Comment:
It is not a typical S-TIR func, which is expected to be:
```python 
@T.prim_func
 def func(A: T.Buffer[(16,), "int32"]):
    for i in T.serial(16):
        with T.block('compute'):
            vi = T.axis.S(16, i)
            A[vi] = vi
```

Here are mistakes in most cases in this RFC:
1. Usually loops are outside the block, and one block only contains a single stmt
2. The `if-else` branch is represented by `T.where` if it's inside a block
3. Please explicitly write block vars, since it's important during schedule.

Reply:
Thank you, and I can update the examples accordingly, though it will be a bit before I have the availability to do so.  (I had been primarily focusing on the algebraic manipulations during the initial drafting.)